File: src/leetCode/c++/Solution97.cpp
//2D-DP O(mn) O(mn)
class Solution97 {
public:
    bool isInterleave(string s1, string s2, string s3) {
        int m = s1.length(), n = s2.length();
        if(m + n != s3.length()) return false;

        vector<vector<bool>> dp(m + 1, vector<bool>(n + 1));
        dp[0][0] = true;
        for(int i=1; i<=m; ++i) dp[i][0] = dp[i - 1][0] && s1[i - 1] == s3[i - 1];
        for(int j=1; j<=n; ++j) dp[0][j] = dp[0][j - 1] && s2[j - 1] == s3[j - 1];

        for(int i=1; i<=m; ++i){
            for(int j=1; j<=n; ++j){//s1或s2的前綴為true且s1或s2也配對到字母時
                dp[i][j] =  (dp[i - 1][j] && s1[i - 1] == s3[i + j - 1])
                         || (dp[i][j - 1] && s2[j - 1] == s3[i + j - 1]);
            }
        }
        return dp[m][n];
    }
};
/* s1 = "aac", s2 = "dbb", s3 = "aadbbc"

        d   b   b
    T   F   F   F
a   T   F   F   F
a   T   T   T   T
c   F   F   F   T

*/

File: java/Solution97.java
package leetCode.java;

//2D-DP O(mn) O(mn)
class Solution97 {
    public boolean isInterleave(String s1, String s2, String s3) {
        int m = s1.length(), n = s2.length();
        if (m + n != s3.length()) {
            return false;
        }

        boolean[][] dp = new boolean[m + 1][n + 1];
        dp[0][0] = true;
        for (int i = 0; i <= m; ++i) {
            for (int j = 0; j <= n; ++j) {
                if (i > 0) {
                    dp[i][j] |= s1.charAt(i - 1) == s3.charAt(i + j - 1) && dp[i - 1][j];
                }
                if (j > 0) {
                    dp[i][j] |= s2.charAt(j - 1) == s3.charAt(i + j - 1) && dp[i][j - 1];
                }
            }   
        }
        return dp[m][n];
    }
}
/*

dp[i][j]代表:
s1 前 i 個字元 + s2 前 j 個字元, 能不能組成 s3 前 i+j 個字元

dp[0][0] = true, 因為 空的 s1 + 空的 s2 = 空的 s3

dp[i][j] =
    (s1.charAt(i - 1) == s3.charAt(i + j - 1) && dp[i - 1][j])
    ||
    (s2.charAt(j - 1) == s3.charAt(i + j - 1) && dp[i][j - 1])

ex: s1 = "aac", s2 = "dbb", s3 = "aadbbc"

        j   0   1   2   3
                d   b   b
i
0           T   F   F   F
1   a       T   F   F   F
2   a       T   T   T   T
3   c       F   F   F   T

res=true
*/
